### PR TITLE
Introduce AMReX-Hydro

### PR DESCRIPTION
**Pull request by ashesh2512**
With recent changes in AMReX, the mac and nodal projectors have been moved from AMReX to AMReX-Hydro. This pull request updates amr-wind to use the latest version of AMReX including introducing cmake dependencies that account for AMReX-Hydro.

### CODE REVIEW

File: cmake/amr-wind-utils.cmake
Comment (jrood-nrel) [inline cmake/amr-wind-utils.cmake:78]: I think this can be removed since it happens in the options just before it.

File: CMakeLists.txt
Comment (jrood-nrel) [inline CMakeLists.txt:208]: Can you remove these accidental spaces?

**Review by jrood-nrel — APPROVED** [review]

File: CMakeLists.txt
Comment (ashesh2512) [reply CMakeLists.txt]: @jrood-nrel I have removed the spaces here - https://github.com/Exawind/amr-wind/pull/551/commits/6d335317b88b0152e4b1e970a886226d4e92b1fa

File: cmake/AMR-WindConfig.cmake.in
Comment (sayerhs) [inline cmake/AMR-WindConfig.cmake.in:17]: Why do we need this? How will this work if we want to use an external `AMReX-hydro` install?

File: CMakeLists.txt
Comment (sayerhs) [inline CMakeLists.txt:192]: Is there no qualified name similar to `AMReX::amrex` for the hydro libraries?

File: cmake/amr-wind-utils.cmake
Comment (sayerhs) [inline cmake/amr-wind-utils.cmake:75]: What about support for external library?

**Review by sayerhs — CHANGES_REQUESTED** [review]

File: cmake/amr-wind-utils.cmake
Comment (ashesh2512) [reply cmake/amr-wind-utils.cmake]: @jrood-nrel Good catch! I have removed this here - https://github.com/Exawind/amr-wind/pull/551/commits/dd621eb635d727fe286ba8e5b8cb8bb6d5a624fe

File: CMakeLists.txt
Comment (ashesh2512) [reply CMakeLists.txt:192]: @sayerhs I looked into it previously, and unless I missed it, unfortunately the hydro libraries don't appear to have a qualified name. Incflo seems to be using `amrex_hydro` too - https://github.com/AMReX-Codes/incflo/blob/development/CMakeLists.txt#L147

Comment (sayerhs) [reply CMakeLists.txt:192]: May be something to request in AMReX-hydro package for future.

File: cmake/AMR-WindConfig.cmake.in
Comment (sayerhs) [reply cmake/AMR-WindConfig.cmake.in:17]: Remove this line as we don't want the following invocation to fail. 

```
find_package(AMR-Wind REQUIRED)
```

File: cmake/amr-wind-utils.cmake
Comment (sayerhs) [reply cmake/amr-wind-utils.cmake:75]: Upon further investigation, AMReX-hydro does not provide install rules and is configured as an `object` library only. As such it can only be used as a submodule within AMR-Wind.

File: cmake/set_amrex_hydro_options.cmake
Comment (sayerhs) [inline cmake/set_amrex_hydro_options.cmake:4]: We don't support EB in AMR-Wind

File: cmake/set_amrex_hydro_options.cmake
Comment (sayerhs) [inline cmake/set_amrex_hydro_options.cmake:16]: These are just being forwarded to AMReX within `hydro` and we are already configuring AMReX before hydro. So these lines are unnecessary.

File: cmake/set_amrex_hydro_options.cmake
Comment (sayerhs) [inline cmake/set_amrex_hydro_options.cmake:3]: Reuse `AMReX_SPACEDIM` here to be consistent, especially with external AMReX (e.g., spack builds)

**Review by sayerhs — CHANGES_REQUESTED** [review]

File: cmake/AMR-WindConfig.cmake.in
Comment (ashesh2512) [reply cmake/AMR-WindConfig.cmake.in]: @sayerhs fixed it here https://github.com/Exawind/amr-wind/pull/551/commits/757c13a898e30702325e4b4f4182bc08be4bc0e6

File: CMakeLists.txt
Comment (sayerhs) [reply CMakeLists.txt:192]: @ashesh2512 Since `amrex_hydro` is an OBJECT library, I believe this should not be present. Have  you confirmed using this _installed_ `amr-wind` library in the exawind driver project?

Comment (ashesh2512) [reply CMakeLists.txt:192]: @sayerhs I have not built the exawind driver using the changes here, although without amrex_hydro as a target, the cmake process fails with the following errors - 

CMake Error: install(EXPORT "AMR-WindTargets" ...) includes target "amrwind_api" which requires target "amrex_hydro" that is not in any export set.
CMake Error: install(EXPORT "AMR-WindTargets" ...) includes target "amrwind_obj" which requires target "amrex_hydro" that is not in any export set.

Comment (sayerhs) [reply CMakeLists.txt:192]: Ok. Thanks for confirming

**Review by sayerhs — APPROVED** [review]

**Comment by ashesh2512** [comment]
Converting to draft to make sure this compiles with exawind-driver

File: amr-wind/CMakeLists.txt
Comment (sayerhs) [inline amr-wind/CMakeLists.txt:44]:
```suggestion
target_link_libraries(${amr_wind_exe_name} PRIVATE ${amr_wind_lib_name})
target_link_libraries(${aw_api_lib} PUBLIC ${amr_wind_lib_name})
```

These targets should inherit `amrex_hydro` from the `${amr_wind_lib_name}` target. So we don't need to add them here.

File: tools/utilities/refine-chkpt/CMakeLists.txt
Comment (sayerhs) [inline tools/utilities/refine-chkpt/CMakeLists.txt:11]: Same here. These should inherit hydro like base `amrex`. Is that not happening automatically?

```suggestion
target_link_libraries(${tool_exe_name} PUBLIC ${amr_wind_lib_name})
```

File: unit_tests/CMakeLists.txt
Comment (sayerhs) [inline unit_tests/CMakeLists.txt:30]:
```suggestion
target_link_libraries(${amr_wind_unit_test_exe_name} PUBLIC ${amr_wind_lib_name})
```

**Review by sayerhs — APPROVED** [review]
Please don't merge until amrex-codes/amrex-hydro#55 is merged.

File: tools/utilities/refine-chkpt/CMakeLists.txt
Comment (jrood-nrel) [reply tools/utilities/refine-chkpt/CMakeLists.txt:11]: It misses symbols without these. I couldn't figure out how to get the inheritance correct. I'm wondering if it's because of the object library.

**Comment by sayerhs** [comment]
@ashesh2512 @jrood-nrel  Ann just merged amrex-codes/amrex-hydro#55 the CI tests should build properly once the submodule is updated to the latest hydro commit.

**Comment by ashesh2512** [comment]
@jrood-nrel Following the merge of https://github.com/AMReX-Codes/AMReX-Hydro/pull/55 I have updated this PR to point to latest AMReX-Hydro. It should be ready to merge at this point.

**Pull request merged**
